runtime/lua/vim/lsp/util.lua

```diff
@@ -423,27 +423,58 @@ function M.make_floating_popup_options(width, height, opts)
   }
 end
 
+local function is_new_item(item)
+  local stack = gettagstack()
+
+  -- Check if we're at the bottom of the tagstack.
+  if stack.curidx <= 1 then return true end
+
+  local top_item = stack.items[stack.curidx-1]
+
+  -- Check if the item at the top of the tagstack is exactly the
+  -- same as the one we want to push.
+  if top_item.tagname ~= item.tagname then return true end
+  for i, v in ipairs(top_item.from) do
+    if v ~= item.from[i] then return true end
+  end
+  return false
+end
+
+local function add_tagstack(item)
+  -- Check whether item is not already at the top of the tagstack.
+  if not is_new_item(item) then return end
+  local winid = vim.fn.win_getid()
+  local items = {item}
+  vim.fn.settagstack(winid, {items=items}, 't')
+end
```

Comment:
I disagree with the first part (no surprise); If the goal of nvim LSP is to have a bare-bones but hackable implementation, then functions should be as small and composable as possible -- otherwise "hacking" means "copying the whole whole file and changing a line".

I do agree about the more expressive name, though!

Reply:
I don't agree with this 😁 I think it's fine to write a function outside of the function, for clarity sake. This function gets redefined every time we call the function, which I don't think makes sense.

I left a comment to either just inline it, or we can move it back out. The name could be clearer, or added w/in a `do` block to hide it's scope, but I am not a big fan of this style choice. Of course, if I'm outnumbered, I will acquiesce 😆